lxd/device/nic_bridged.go

```diff
@@ -1092,6 +1087,59 @@ func (d *nicBridged) setFilters() (err error) {
 	return nil
 }
 
+func (d *nicBridged) allowedNetworks(IPv4 net.IP, IPv6 net.IP) ([]*net.IPNet, error) {
+	// If no allocated IPv4 address for filtering and filtering enabled, then block all IPv4 traffic.
+	var allowedIPv4Nets []*net.IPNet
+	if shared.IsTrue(d.config["security.ipv4_filtering"]) {
+		if IPv4 != nil {
+			allowedIPv4Nets = append(allowedIPv4Nets, &net.IPNet{IP: IPv4, Mask: net.IPMask{255, 255, 255, 255}})
+		}
+
+		routes := append(util.SplitNTrimSpace(d.config["ipv4.routes"], ",", -1, true),
+			util.SplitNTrimSpace(d.config["ipv4.routes.external"], ",", -1, true)...)
+
+		for _, route := range routes {
+			ipNet, err := network.ParseIPCIDRToNet(route)
+			if err != nil {
+				return nil, err
+			}
+			allowedIPv4Nets = append(allowedIPv4Nets, ipNet)
+		}
+
+		if len(allowedIPv4Nets) == 0 {
+			allIPv4Addresses, _ := network.ParseIPCIDRToNet(firewallDrivers.FilterIPv4NetAll)
+			allowedIPv4Nets = append(allowedIPv4Nets, allIPv4Addresses)
+		}
+	}
+
+	// If no allocated IPv6 address for filtering and filtering enabled, then block all IPv6 traffic.
+	var allowedIPv6Nets []*net.IPNet
+	if shared.IsTrue(d.config["security.ipv6_filtering"]) {
+		if IPv6 != nil {
+			allowedIPv6Nets = append(allowedIPv6Nets, &net.IPNet{IP: IPv6, Mask: net.IPMask{255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255}})
+		}
+
+		routes := append(util.SplitNTrimSpace(d.config["ipv6.routes"], ",", -1, true),
```

Comment:
Parsing the routes here from the config directly rather than accepting them either as a separate allowedRoutes argument (or merging them with the existing IP arguments into a single allowedIPs argument) would seem to blur the separate of concerns of this function because it accepts individual parsed IPs from config, and yet also parses some of them IPs directly from config.

Aside from the mixture in separation of concerns, this may also present a problem when updating the NIC config, because the old filters need to be removed using the previous version of the config (which is provided to the Update function as a separate argument) and so anything to do with "updating" config needs to be able to accept a config set to operate on without directly accessing `d.config` (which is always the current config).

Reply:
Take a look at the use of `d.removeFilters` to see what I mean in this case.

Reply:
I've removed the `nicBridged` receiver from this function and accepted a `deviceConfig.Device` as an argument to ensure that all the IPs come from a single devices configuration (or old configuration).